File: database_integration/enhance_db/src/bin/06_restructure_nationalities.rs
/// Restructure nationalities table:
/// - Put wikidata_id first
/// - Add en_wikipedia_url column (from extracted sitelinks)
/// - Add lat/lon location columns (from extracted locations)
/// - Add modern_country_name column
use anyhow::Result;
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::io::Write;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const SITELINKS_PATH: &str = "data/all_humans/nationality_sitelinks.json";
const LOCATIONS_PATH: &str = "data/all_humans/nationality_locations.json";
const NAT_COUNTRIES_PATH: &str = "data/all_humans/nationality_countries.json";
const TASK_LOG: &str = "task.log";

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

fn main() -> Result<()> {
    log("[DB] 06: Restructuring nationalities table...");

    // Load sitelinks
    let sitelinks_json = fs::read_to_string(SITELINKS_PATH)?;
    let sitelinks: HashMap<String, Value> = serde_json::from_str(&sitelinks_json)?;
    log(&format!("[DB] Loaded {} nationality sitelinks", sitelinks.len()));

    // Load locations
    let locations_json = fs::read_to_string(LOCATIONS_PATH)?;
    let locations: HashMap<String, Value> = serde_json::from_str(&locations_json)?;
    log(&format!("[DB] Loaded {} nationality locations", locations.len()));

    // Load nationality -> country mapping
    let nat_countries_json = fs::read_to_string(NAT_COUNTRIES_PATH)?;
    let nat_countries: HashMap<String, Value> = serde_json::from_str(&nat_countries_json)?;
    log(&format!("[DB] Loaded {} nationality->country mappings", nat_countries.len()));

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;")?;

    // Read existing nationalities data
    let mut existing: Vec<(String, i64, String, Option<String>, Option<String>)> = Vec::new();
    {
        let mut stmt = conn.prepare("SELECT name_en, count, wikidata_id, description_en, instance_of FROM nationalities")?;
        let rows = stmt.query_map([], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, i64>(1)?,
                row.get::<_, String>(2).unwrap_or_default(),
                row.get::<_, Option<String>>(3)?,
                row.get::<_, Option<String>>(4)?,
            ))
        })?;
        for r in rows {
            existing.push(r?);
        }
    }
    log(&format!("[DB] Read {} existing nationalities", existing.len()));

    // Drop and recreate with wikidata_id first
    conn.execute_batch("DROP TABLE IF EXISTS nationalities_backup;")?;
    conn.execute_batch("ALTER TABLE nationalities RENAME TO nationalities_backup;")?;

    conn.execute_batch(
        "CREATE TABLE nationalities (
            wikidata_id TEXT PRIMARY KEY,
            name_en TEXT,
            count INTEGER DEFAULT 0,
            description_en TEXT,
            instance_of TEXT,
            en_wikipedia_url TEXT,
            lat REAL,
            lon REAL,
            modern_country_name TEXT
        );"
    )?;

    let mut stmt = conn.prepare(
        "INSERT OR IGNORE INTO nationalities (wikidata_id, name_en, count, description_en, instance_of, en_wikipedia_url, lat, lon, modern_country_name)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)"
    )?;

    let pb = ProgressBar::new(existing.len() as u64);
    pb.set_style(ProgressStyle::default_bar()
        .template("{msg} [{bar:40}] {pos}/{len} ({eta})")
        .unwrap());
    pb.set_message("Inserting nationalities");

    for (name_en, count, wikidata_id, description, instance_of) in &existing {
        let wiki_url = sitelinks.get(wikidata_id).and_then(|v| v.as_str()).map(String::from);

        let (lat, lon) = if let Some(loc) = locations.get(wikidata_id) {
            (
                loc.get("lat").and_then(|v| v.as_f64()),
                loc.get("lon").and_then(|v| v.as_f64()),
            )
        } else {
            (None, None)
        };

        // Get modern_country_name
        let modern_country = nat_countries.get(wikidata_id).and_then(|v| {
            v.get("country_name").and_then(|cn| cn.as_str()).map(String::from)
        });

        stmt.execute(params![
            wikidata_id,
            name_en,
            count,
            description,
            instance_of,
            wiki_url,
            lat,
            lon,
            modern_country,
        ])?;
        pb.inc(1);
    }
    pb.finish();

    conn.execute_batch("DROP TABLE IF EXISTS nationalities_backup;")?;

    // Create indexes
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_nationalities_name ON nationalities(name_en);")?;
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_nationalities_country ON nationalities(modern_country_name);")?;

    log("[DB] 06: Done. Nationalities restructured with wikidata_id first, sitelinks, locations, and country.");
    Ok(())
}

File: scripts/database_integration_scripts_V2/legacy/06_restructure_nationalities.py
"""06 - Restructure the nationalities table.

Mirrors `enhance_db/src/bin/06_restructure_nationalities.rs`.

  Inputs : data/all_humans/nationality_sitelinks.json   {qid: en_url}
           data/all_humans/nationality_locations.json   {qid: {lat, lon}}
           data/all_humans/nationality_countries.json   {qid: {country_name}}
           nationalities (existing rows)
  Output : nationalities (rebuilt) with wikidata_id PK first, plus
           en_wikipedia_url, lat/lon, modern_country_name columns and
           two indexes.

Usage
-----
    python3 06_restructure_nationalities.py
    python3 06_restructure_nationalities.py --full
"""

from __future__ import annotations

import json
import sqlite3
import tempfile
from pathlib import Path

from common import (
    ALL_HUMANS_DIR,
    insert_rows,
    load_json,
    log,
    open_db,
    parse_run_mode,
)

SITELINKS_PATH = ALL_HUMANS_DIR / "nationality_sitelinks.json"
LOCATIONS_PATH = ALL_HUMANS_DIR / "nationality_locations.json"
NAT_COUNTRIES_PATH = ALL_HUMANS_DIR / "nationality_countries.json"


def run(
    conn: sqlite3.Connection,
    sitelinks_path: Path = SITELINKS_PATH,
    locations_path: Path = LOCATIONS_PATH,
    nat_countries_path: Path = NAT_COUNTRIES_PATH,
) -> int:
    log("[DB] 06: Restructuring nationalities table...")

    sitelinks = load_json(sitelinks_path)
    log(f"[DB] Loaded {len(sitelinks)} nationality sitelinks")
    locations = load_json(locations_path)
    log(f"[DB] Loaded {len(locations)} nationality locations")
    nat_countries = load_json(nat_countries_path)
    log(f"[DB] Loaded {len(nat_countries)} nationality->country mappings")

    existing = conn.execute(
        "SELECT name_en, count, wikidata_id, description_en, instance_of FROM nationalities"
    ).fetchall()
    log(f"[DB] Read {len(existing)} existing nationalities")

    conn.execute("DROP TABLE IF EXISTS nationalities_backup")
    conn.execute("ALTER TABLE nationalities RENAME TO nationalities_backup")
    conn.execute(
        """
        CREATE TABLE nationalities (
            wikidata_id TEXT PRIMARY KEY,
            name_en TEXT,
            count INTEGER DEFAULT 0,
            description_en TEXT,
            instance_of TEXT,
            en_wikipedia_url TEXT,
            lat REAL,
            lon REAL,
            modern_country_name TEXT
        )
        """
    )

    insert_sql = (
        "INSERT OR IGNORE INTO nationalities "
        "(wikidata_id, name_en, count, description_en, instance_of, "
        "en_wikipedia_url, lat, lon, modern_country_name) "
        "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)"
    )
    rows: list[tuple] = []
    for name_en, count, wid, description, instance_of in existing:
        wid = wid or ""
        wiki_url = sitelinks.get(wid)
        if isinstance(wiki_url, dict):
            wiki_url = wiki_url.get("url") or wiki_url.get("en")
        loc = locations.get(wid) or {}
        lat = loc.get("lat") if isinstance(loc, dict) else None
        lon = loc.get("lon") if isinstance(loc, dict) else None
        nat = nat_countries.get(wid) or {}
        modern_country = nat.get("country_name") if isinstance(nat, dict) else None
        rows.append((wid, name_en, count, description, instance_of,
                     wiki_url, lat, lon, modern_country))

    try:
        from tqdm import tqdm
        rows_iter = tqdm(rows, desc="Inserting nationalities", unit="row")
    except ImportError:
        rows_iter = rows
    conn.executemany(insert_sql, rows_iter)
    conn.commit()

    conn.execute("DROP TABLE IF EXISTS nationalities_backup")
    conn.execute("CREATE INDEX IF NOT EXISTS idx_nationalities_name ON nationalities(name_en)")
    conn.execute("CREATE INDEX IF NOT EXISTS idx_nationalities_country ON nationalities(modern_country_name)")
    conn.commit()

    log("[DB] 06: Done. Nationalities restructured.")
    return len(rows)


def _sample_main() -> None:
    fake_sitelinks = {"Q142": "https://en.wikipedia.org/wiki/France"}
    fake_locations = {"Q142": {"lat": 46.0, "lon": 2.0}}
    fake_countries = {"Q142": {"country_name": "France"}}

    with tempfile.TemporaryDirectory() as tmp:
        sl_path = Path(tmp) / "sl.json"
        loc_path = Path(tmp) / "loc.json"
        ncp_path = Path(tmp) / "ncp.json"
        sl_path.write_text(json.dumps(fake_sitelinks))
        loc_path.write_text(json.dumps(fake_locations))
        ncp_path.write_text(json.dumps(fake_countries))

        db_path = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db_path) as seed:
            seed.execute(
                "CREATE TABLE nationalities ("
                "name_en TEXT PRIMARY KEY, count INTEGER, wikidata_id TEXT, "
                "description_en TEXT, instance_of TEXT)"
            )
            insert_rows(seed, "nationalities", [
                {"name_en": "French", "count": 1234, "wikidata_id": "Q142",
                 "description_en": "from France", "instance_of": "country"},
                {"name_en": "Atlantean", "count": 0, "wikidata_id": "Q999",
                 "description_en": None, "instance_of": None},
            ])
            seed.commit()

        with open_db(db_path) as conn:
            n = run(conn, sl_path, loc_path, ncp_path)
            rows = conn.execute(
                "SELECT wikidata_id, name_en, count, en_wikipedia_url, lat, lon, modern_country_name "
                "FROM nationalities ORDER BY count DESC"
            ).fetchall()
        log(f"[sample] inserted {n} nationalities")
        for r in rows:
            log(f"  nationalities: {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()
